### DISCUSSION TITLE
Welcome to arrow-udf Discussions!

### DISCUSSION THREAD

**Discussion by Xuanwo** [risingwavelabs/arrow-udf#66] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by viethqb** [comment]
As a Data Platform Engineer, I'm very fond of versatile database solutions like Trino, Doris, Starocks, and RisingWave. These tools make problem-solving quicker and more efficient.

Trino, in particular, lets me easily integrate with various data sources, making data ingestion seamless. When coupled with dbt and astronomer-cosmos, I can centrally manage the entire ETL process, which is very convenient.

However, for data sources like APIs, writing UDFs in Java for Trino is quite time-consuming and laborious. After exploring the combination of Arrow-UDF and RisingWave, I had a new idea:
- Write a UDF to fetch data from the API for RisingWave.
- Use Trino Table Functions to call this UDF.
- This way, I can perform all necessary operations using just Trino.

During implementation, I faced some challenges:

- Native UDF: Couldn't import libraries like requests, httpx for data fetching.
- Python Remote UDF: Meets requirements but only uses one core. I can address this by increasing replicas when deploying on Kubernetes.
- Java Remote UDF: I dislike and want to avoid Java and JVM languages.

	
Though successful, I believe the approach of Trino calling RisingWave, and then RisingWave calling the UDF server is suboptimal, increasing network overhead.

I'm seeking a versatile data processing tool that inherits and combines the best aspects of existing solutions like Trino, Doris, Starocks, and RisingWave. This ideal tool should meet the following requirements:

Flexible data ingestion (Trino, RisingWave):
- Supports both real-time and batch data ingestion.
- Easily write user-defined functions (UDFs) to integrate with non-standard data sources. (Snowflake, arrow-udf)

User-friendly query interface (Doris, Starocks, and RisingWave):
- Provides a familiar query interface like MySQL or PostgreSQL for easy user interaction.

Ability to interact with open data formats (Trino, Doris, and Starocks):
- Supports reading and writing data on popular data lakehouse platforms like Iceberg, Hudi, Delta, Paimon.

Diverse data processing (Trino, Doris, Starocks, and RisingWave):
- Supports both batch and real-time data processing.

Easily deliver data to other systems (Risingwave).

Separated architecture (Starocks, and RisingWave):
- Separates storage and computing for improved performance and scalability.

Programming language:
- Doesn't use JVM languages. Rust is a potential choice due to its high performance and reliability.
